### PR TITLE
MCR-3646 Harmonize edge cases and improve error and log messages of annotation based configuration

### PR DESCRIPTION
**Pull request by toKrause**
[Link to jira](https://mycore.atlassian.net/browse/MCR-3646).

# Pull Request Checklist (Author)

Please go through the following checklist before assigning the PR for review:

## Ticket & Documentation
- [x] The issue in the ticket is clearly described and the solution is documented.
- [x] Design decisions (if any) are explained.
- [x] The ticket references the correct source and target branches.
- [x] The `fixed-version` is correctly set in the ticket and matches the PR's target branch (`main`).

## Feature & Improvement Specific Checks
- [x] Instructions on how to test or use the feature are included or linked (e.g. to documentation).
- [ ] For UI changes: before & after screenshots are attached.
- [ ] New features or migrations are documented.
- [ ] Does this change affect existing applications, data, or configurations?
  - [x] Yes: Is a migration required? If yes, describe it.
  - [ ] Breaking change is marked in the **commit message**.

Behavior of some edge cases changes. See ticket for complete list. Migration guide follows.

## Testing
- [x] I have tested the changes locally.
- [x] The feature behaves as described in the ticket.
- [x] Were existing tests modified?
  - [x] Yes: explain the changes for reviewers.

See ticket. Code contains exhaustive list of tests cases.

## MCR Conventions & Metadata
- [x] [MCR naming conventions](https://www.mycore.de/documentation/developer/conventions/) are followed
- [ ] If the **public API** has changed:
  - [ ] Old API is deprecated or a migration is documented.
  - [ ] If not, no action needed.
- [x] Java license headers are added where necessary.
- [x] Javadoc is written for non-self-explanatory classes/methods (Clean Code).
- [x] All configuration options are documented in Javadoc and `mycore.properties`.
- [x] No default properties are hardcoded — all set via `mycore.properties`.

## Multi-Repo Considerations
- [ ] Is an equivalent PR in `MIR` required?
  - [ ] If yes, is it already created?

### CODE REVIEW

**Review by sebhofmann — CHANGES_REQUESTED** [review]
The 9600 lines across these six files aren't really maintainable. They're copies of the same structure, walking the truth table cell by cell. When the rules need to change, nobody's going to want to touch this.

The spec is ~10 rules plus 3 log triggers, driven by the same four input axes. Could we encode those rules once in the test as expectedOutcome(required, classProp, sentinel, valueProp) returning the expected instance and logs, then loop over all input combinations? A rule change is one edit, not 540. The error message structure (type, property, class, field, reason) can be a single helper applied at each exception site.

If that feels too much, the lighter version is `@ParameterizedTest` walking the same combinations. That needs the test setup to move off `@MCRTestConfiguration` / `@MCRTestProperty` though, since annotations can't take runtime parameters. Properties would be set programmatically in `@BeforeEach` (or per case) from the arguments.

**Comment by toKrause** [comment]
I'm not sure how practical parameterization would be, since different test cases use differently annotated configured classes, and I didn't want to develop a complete testing framework. It was much simpler to test each case individually than to invent something clever.

**Comment by toKrause** [comment]
... but I'll look into it.

**Comment by toKrause** [comment]
I've reworked one class as an example: https://github.com/MyCoRe-Org/mycore/pull/2861/changes#diff-8797c2630ef7a30469903ddb89e83200e10c8891a11c7b85c47ae6e5b2e7a581

**Comment by toKrause** [comment]
If this style is okay for you, I'll rework the rest accordingly.

**Comment by sebhofmann** [comment]
Yes that style is okay for me. Just notices some typos shoud -> should.

File: mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java [hunk 1]
```diff
@@ -229,19 +234,96 @@ private static String targetTypeName(Target<?> target) {
...
+        if (Objects.equals("", annotationName)) {
+            return configuration.name().canonical();
+        } else {
+            return configuration.name().canonical() + "." + annotationName;
+        }
+    }
+
+    private static MCRConfigurationException incompatibilityException(String property, Target<?> target,
+        Class<?> annotationValueClass, Object instance) {
+        throw new MCRConfigurationException("Instance of class " + instance.getClass().getName()
```
Comment (sebhofmann) [inline mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java]: This helper method should return the Exception or have void as return type. I is also inconsistent compared to `nonIntegerKeyException` and `inconsistentKeysException`

File: mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java [hunk 2]
```diff
@@ -229,19 +234,96 @@ private static String targetTypeName(Target<?> target) {
...
+            + "', configured in " + property + ", is incompatible with annotation value class "
+            + annotationValueClass.getName() + " for target " + targetTypeName(target)
+            + " '" + target.name() + "' in configured class " + target.declaringClass().getName());
+    }
+
+    private static MCRConfigurationException missingException(String property, Target<?> target,
+        String description) {
+        throw new MCRConfigurationException(
```
Comment (sebhofmann) [inline mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java]: This helper method should return the Exception or have void as return type. I is also inconsistent compared to `nonIntegerKeyException` and `inconsistentKeysException`

File: mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java [hunk 3]
```diff
@@ -229,19 +234,96 @@ private static String targetTypeName(Target<?> target) {
...
+            capitalize(description) + ", configured in " + property + " (and its sub-properties)," +
+                " for target " + targetTypeName(target) + " '" + target.name() + "' in configured class "
+                + target.declaringClass().getName()
+                + " is missing");
+    }
+
+    private static MCRConfigurationException emptyException(String property, Target<?> target, String description) {
+        throw new MCRConfigurationException(
```
Comment (sebhofmann) [inline mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java]: This helper method should return the Exception or have void as return type. I is also inconsistent compared to `nonIntegerKeyException` and `inconsistentKeysException`

File: mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java [hunk 4]
```diff
@@ -229,19 +234,96 @@ private static String targetTypeName(Target<?> target) {
...
+            capitalize(description) + ", configured in " + property + " (and its sub-properties)," +
+                " for target " + targetTypeName(target) + " '" + target.name() + "' in configured class "
+                + target.declaringClass().getName()
+                + " is empty");
+    }
+
+    private static MCRConfigurationException emptyRawException(String property, Target<?> target, String description) {
+        throw new MCRConfigurationException(
```
Comment (sebhofmann) [inline mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java]: This helper method should return the Exception or have void as return type. I is also inconsistent compared to `nonIntegerKeyException` and `inconsistentKeysException`

File: mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java [hunk 5]
```diff
@@ -1299,7 +1449,59 @@ public Map<String, String> get(MCRInstanceConfiguration configuration, Target<?>
...
+
+        protected Object getInstance(String property, Target<?> target, Class<?> valueClass,
+            MCRInstanceConfiguration nestedConfiguration, MCRSentinel sentinel) {
+
+            Set<Option> options = createOptions(valueClass);
+            boolean implicitValueClass = Modifier.isFinal(valueClass.getModifiers());
+
+            if (sentinel.enabled()) {
+                boolean sentinelValue = sentinel.defaultValue();
+                String configuredSentinelValue = nestedConfiguration.properties().remove(sentinel.name());
```
Comment (sebhofmann) [inline mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java:1463]: This modifies the the `nestedConfiguration.properties()`, does this get evaluated just once? Maybe better use `properties().get(sentinel.name())`.

File: mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java [hunk 6]
```diff
@@ -229,19 +234,96 @@ private static String targetTypeName(Target<?> target) {
...
+                + target.declaringClass().getName()
+                + " is empty");
+    }
+
+    private static MCRConfigurationException nonIntegerKeyException(String property, Target<?> target,
+        String key, String description, NumberFormatException exception) {
+        return new MCRConfigurationException(
+            capitalize(description) + ", configured in " + property + " (and its sub-properties)," +
+                " for target " + targetTypeName(target) + " '" + target.name() + "' in configured class "
+                + target.declaringClass().getName() + " is has element with non-integer key " + key,
```
Comment (sebhofmann) [inline mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java]: Typo: "is has element with non-integer key" -> "has element with non-integer key".

File: mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java [hunk 7]
```diff
@@ -901,29 +983,70 @@ public String get(MCRInstanceConfiguration configuration, Target<?> target) {
...
+
+            if (sentinel.enabled()) {
+                boolean sentinelValue = sentinel.defaultValue();
+                String configurationValue = properties.get(prefix + "." + sentinel.name());
+                if (configurationValue != null) {
+                    sentinelValue = Boolean.parseBoolean(configurationValue);
+                }
+                if (sentinelValue == sentinel.rejectionValue()) {
+                    if (LOGGER.isInfoEnabled()) {
+                        LOGGER.info("[SENTINEL] Ignoring {} {} and all sup-properties",
```
Comment (sebhofmann) [inline mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java]: Typo: sub-properties

File: mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java [hunk 8]
```diff
@@ -169,25 +167,32 @@ public static <S> S getInstance(Class<S> superClass, MCRInstanceConfiguration co
...
         }
         Class<S> targetClass = getClass(configuration.name().actual(), configuration.className());
-        Object instance = createInstanceDirectorViaProxy(targetClass, configuration);
+        Object instance = createInstanceDirectOrViaProxy(targetClass, configuration);
         if (superClass.isAssignableFrom(instance.getClass())) {
             return superClass.cast(instance);
         } else {
-            throw new MCRConfigurationException("Configured instance of class " + instance.getClass().getName()
+            throw new MCRConfigurationException("Instance of class " + instance.getClass().getName()
                 + " is incompatible with intended super class " + superClass.getName() + "'"
```
Comment (sebhofmann) [inline mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java]: Missing opening `'`

File: mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java (hunk 1 above)
Comment (sebhofmann) [reply mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java]: Missing opening `'`

**Review by sebhofmann — CHANGES_REQUESTED** [review]

File: mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java (hunk 5 above)
Comment (toKrause) [reply mycore-base/src/main/java/org/mycore/common/config/MCRConfigurableInstanceHelper.java:1463]: The configuration is only used for one instantiation process. it is not reused and can therefore safely be modified. The idea here is to remove the sentinel value, so it isn't available anymore as a configurable value (same as with the `Class` property, which is also not part of `properties()`).

This is intended to prevent a situation where one property influences two things. Example:

```java
    @MCRInstance(name = FOOBAR_KEY, valueClass = FooBar.class, required = false, sentinel = @MCRSentinel(name = "Active"))
    public Foobar foobar;
    
    public static class Foobar {
        
        @MCRProperty(name = "Class")
        private String classString;
        
        @MCRProperty(name = "Active")
        private String activeString;
        
    }
```

Both of these properties should not be used as a configuration value for the instance, because they are already used as as a configuration value for the instantiation process.

**Review by sebhofmann — APPROVED** [review]
